discovery-provider/src/eth_indexing/event_scanner.py

```diff
@@ -0,0 +1,434 @@
+        # self.state.restore()
+
+    def restore(self):
+        """Restore the last scan state from redis."""
+        restored = redis_get_or_restore(self.redis, eth_indexing_last_scanned_block_key)
+        self.last_scanned_block = int(restored) if restored else MIN_SCAN_START_BLOCK
+        logger.info(f"Restored last scanned block ({self.last_scanned_block})")
+
+    def save(self, block_number: int):
+        """Save at the end of each chunk of blocks, so we can resume in the case of a crash or CTRL+C
+        Next time the scanner is started we will resume from this block
+        """
+        self.last_scanned_block = block_number
+        logger.info(f"Saving last scanned block ({self.last_scanned_block}) to redis")
+        redis_set_and_dump(
+            self.redis,
+            eth_indexing_last_scanned_block_key,
+            str(self.last_scanned_block),
+        )
+
+    def get_block_timestamp(self, block_num) -> Union[datetime.datetime, None]:
+        """Get Ethereum block timestamp"""
+        try:
+            block_info = self.web3.eth.getBlock(block_num)
+        except BlockNotFound:
+            # Block was not mined yet,
+            # minor chain reorganisation?
+            return None
+        last_time = block_info["timestamp"]
+        return datetime.datetime.utcfromtimestamp(last_time)
+
+    def get_suggested_scan_end_block(self):
+        """Get the last mined block on Ethereum chain we are following."""
+
+        # Do not scan all the way to the final block, as this
+        # block might not be mined yet
+        return self.web3.eth.blockNumber - ETH_BLOCK_TAIL_OFFSET
+
+    def get_last_scanned_block(self) -> int:
+        """The number of the last block we have stored."""
+        return self.last_scanned_block
+
+    def process_event(self, block_when: datetime.datetime, event: TransferEvent) -> str:
+        """Record a ERC-20 transfer in our database."""
+        # Events are keyed by their transaction hash and log index
+        # One transaction may contain multiple events
+        # and each one of those gets their own log index
+
+        # event_name = event.event # "Transfer"
+        log_index = event["logIndex"]  # Log index within the block
+        # transaction_index = event.transactionIndex  # Transaction index within the block
+        txhash = event["transactionHash"].hex()  # Transaction hash
+        block_number = event["blockNumber"]
+
+        # Convert ERC-20 Transfer event to our internal format
+        args = event["args"]
+        transfer = {
+            "from": args["from"],
+            "to": args["to"],
+            "value": args["value"]
+            / 1e18,  # divide by 10^18 to get correct transfer value
+            "timestamp": block_when,
+        }
+
+        # add user ids from the transfer event into the balance refresh queue
+        transfer_event_wallets = [transfer["from"].lower(), transfer["to"].lower()]
+        with self.db.scoped_session() as session:
```

Comment:
do we want a join here? wouldn't two queries be faster and then a dedupe in a set in memory?